source/s3_request.c

```diff
@@ -95,3 +124,259 @@ static void s_s3_request_destroy(void *user_data) {
 
     aws_mem_release(request->allocator, request);
 }
+
+static void s_s3_request_metrics_destroy(void *arg) {
+    struct aws_s3_request_metrics *metrics = arg;
+    if (metrics == NULL) {
+        return;
+    }
+    aws_http_headers_release(metrics->req_resp_info_metrics.response_headers);
+    aws_byte_buf_clean_up(&metrics->req_resp_info_metrics.request_path_query);
+    aws_byte_buf_clean_up(&metrics->req_resp_info_metrics.host_address);
+    aws_byte_buf_clean_up(&metrics->crt_info_metrics.ip_address);
+
+    aws_mem_release(metrics->allocator, metrics);
+}
+
+struct aws_s3_request_metrics *aws_s3_request_metrics_new(
+    struct aws_allocator *allocator,
+    struct aws_http_message *message) {
+    struct aws_s3_request_metrics *metrics = aws_mem_calloc(allocator, 1, sizeof(struct aws_s3_request_metrics));
+    metrics->allocator = allocator;
+    struct aws_byte_cursor out_path;
+    AWS_ZERO_STRUCT(out_path);
+    int err = aws_http_message_get_request_path(message, &out_path);
+    /* If there is no path of the message, it should be a program error. */
+    AWS_ASSERT(!err);
+    err = aws_byte_buf_init_copy_from_cursor(&metrics->req_resp_info_metrics.request_path_query, allocator, out_path);
+    AWS_ASSERT(!err);
+
+    /* Get the host header value */
+    struct aws_byte_cursor host_header_value;
+    AWS_ZERO_STRUCT(host_header_value);
+    struct aws_http_headers *message_headers = aws_http_message_get_headers(message);
+    AWS_ASSERT(message_headers);
+    err = aws_http_headers_get(message_headers, g_host_header_name, &host_header_value);
+    AWS_ASSERT(!err);
+    err =
+        aws_byte_buf_init_copy_from_cursor(&metrics->req_resp_info_metrics.host_address, allocator, host_header_value);
+    AWS_ASSERT(!err);
+
+    (void)err;
+    aws_ref_count_init(&metrics->ref_count, metrics, s_s3_request_metrics_destroy);
+
+    return metrics;
+}
+struct aws_s3_request_metrics *aws_s3_request_metrics_acquire(struct aws_s3_request_metrics *metrics) {
+    if (!metrics) {
+        return NULL;
+    }
```

Comment:
```suggestion
    AWS_PRECONDITION(metrics);
```

only destroy() and release() functions should tolerate NULL

Reply:
I thought acquire also treats NULL https://github.com/awslabs/aws-c-http/pull/359#discussion_r789280477.

Reply:
Huh, I was unaware of this

a quick ctrl-shift-f shows that acquire() functions' tolerance for NULL is wildly inconsistent throughout the codebase :|

whatever, sure, tolerate NULL I could see it both ways
PRO: consistent with release()
CON: the more functions you have that tolerate NULL, the sketchier your codebase gets, as an unexpected NULL gets blows up much later than it might have otherwise

don't super care to spend the time arguing